### ISSUE TITLE
Multiple improvements to test suite

### ISSUE THREAD

**Issue by nickevansuk** [openactive/openactive-test-suite#34]
1) Improve `flow.C1()`:
- To accept request template file as a parameter, to allow some test suites to customise their request templates
- Note that some tests will book more than one event, so we will need to account for multiple OrderItems (see comment https://github.com/openactive/openactive-test-suite/issues/34#issuecomment-589641493)
- Include validation on the output of C1 with the correct validation mode (and if status code is not `200` then use "OpenBookingError" validation mode)

2) Create something like `state.getRandomEvent(["ScheduledSession", "SecheduledSession", "Slot"])` - note this is named generically with the specified type of event passed through to the microservice, preparing the way for https://github.com/openactive/openactive-test-suite/issues/21 - so we can create a suite of tests that works for random stuff specifically. Note passing an array of event types also allows for the creation of more than one event. In fact - perhaps even better - should such an array actually live in [this file](https://github.com/openactive/openactive-test-suite/blob/enhanced-logging/packages/openactive-integration-tests/test/flows/book-and-cancel/book-and-customer-cancel-success/book-free-test.js#L8) with a `randomEvent: ["ScheduledSession", "SecheduledSession", "Slot"]` array used in place of the `event` array? That way `state.createEvent(testEvent)` (see 4 below) could do both random and explicit modes depending on `testEvent`?

3) Ensure that flow.C1 pattern above extended to all endpoints (https://www.openactive.io/open-booking-api/EditorsDraft/#paths-and-verbs), noting the special treatment of `OrdersFeed` that we already have in the code

4) Ensure that the file https://github.com/openactive/openactive-test-suite/blob/enhanced-logging/packages/openactive-integration-tests/test/flows/book-and-cancel/book-and-customer-cancel-success/book-nonfree-test.js can include different `@type` (e.g. "SchuduledSession") - which it should do without issues. So perhaps we simply rename `createScheduledSession()` to be more generic (e.g. `createEvent()`), as it should handle whichever type is specified in the [file](https://github.com/openactive/openactive-test-suite/blob/enhanced-logging/packages/openactive-integration-tests/test/flows/book-and-cancel/book-and-customer-cancel-success/book-free-test.js#L10).

5) Generalise _some_ of C1, C2 tests etc to provide shared examples that can be reused across shared test suites.

**Comment by nickevansuk** [comment]
@ylt note that I've expanded the above with some additional ideas, and also to include booking more than one event. To make this work within the existing framework, how about:

Within the templates we repeat the  OrderItems within the OrderItems (e.g. see below)? Then pass these through all the way from e.g. an array specified [here](https://github.com/openactive/openactive-test-suite/blob/enhanced-logging/packages/openactive-integration-tests/test/flows/book-and-cancel/book-and-customer-cancel-success/book-free-test.js#L8)?

For example
```
  "orderedItem": [
{{#each orderItem}}
    {
      "@type": "OrderItem",
      "position": {{& position}},
      "acceptedOffer": {
        "@type": "Offer",
        "@id": "{{& offerId}}"
      },
      "orderedItem": {
        "@type": "ScheduledSession",
        "@id": "{{& opportunityId}}"
      }
    }
  {{/each}}
  ],
```

I've also added to `getRandomEvent` notes in the example above

**Comment by ylt** [comment]
Using an approach like this should be much easier than using Mustache: https://github.com/openactive/openactive-test-suite/pull/37

**Comment by nickevansuk** [comment]
Great plan!

**Comment by nickevansuk** [comment]
@ylt just to note that C1 validation doesn't seem to be running in the master branch

**Comment by ylt** [comment]
Interesting, I'll have a look into this. This should be resolved as part of the shared-behaviours in point 5 though :)

**Comment by nickevansuk** [comment]
Sure no rush if more efficient to do as part of point 5 :)

**Comment by nickevansuk** [comment]
> Include validation on the output of C1 with the correct validation mode (and if status code is not 200 then use "OpenBookingError" validation mode)

Although we did say that validation would be included in `flow.C1()` - so assume that it should in fact be part of that?

**Comment by ylt** [comment]
Seems to be running fine in https://github.com/openactive/openactive-test-suite/pull/38 ![openactive-integration-tests (zsh) 2020-02-27 13-22-17](https://user-images.githubusercontent.com/705668/75449578-1eb57980-5965-11ea-9e93-3a1c25e2e0a5.png) :)

**Comment by ylt** [comment]
That error handling is implemented in https://github.com/openactive/openactive-test-suite/pull/36

**Comment by ylt** [comment]
Actually finding much of these fairly complicated to implement. The flow component is only designed to deal with singular bits of info and it's hard to figure out how to adapt it to work with multiple in a nice way.

I'm thinking that the best way of defining this may be to come up with some test cases that would require these, then see how a solution naturally evolves as it's being developed.

**Comment by nickevansuk** [comment]
@ylt funny I was just about to post in here...

C1 is still not validating in master

And if you try pointing it at https://everyoneactivebookingfacade-test.azurewebsites.net/, then many of the endpoints don't validate

Perhaps there's some kind of race condition or similar happening here?

**Comment by nickevansuk** [comment]
note you'll need to turn random on to use the endpoint above

**Comment by nickevansuk** [comment]
Ok so I've got the random opportunity part of this working at a basic level (single OrderItem), in commit https://github.com/openactive/openactive-test-suite/commit/ec96f674efda6682aa13baacaa4c1645a53fad22. See `test/flows/book-only/` for my attempt at this.

This is the specific file: https://github.com/openactive/openactive-test-suite/commit/ec96f674efda6682aa13baacaa4c1645a53fad22#diff-121446740cc34c991fa12b7383997ee6

I think I've also might have identified the cause of one of the "race conditions". It looks like the flows are lacking some defensive programming for cases where a prerequisite call fails - specifically those calls that hit the microservice and wait for the feed - getMatch and getOrder. I've made a quick fix for `getMatch`, but we need something similar for `getOrder`.

If you run `npm test -- test/flows/book-only/` against the reference implementation on master now you'll see the issues for C1 validation - it's still not running. Very strange!
